torch/testing/_internal/compilable_quantized_tensor.py

```diff
@@ -0,0 +1,366 @@
+
+        with torch._C.DisableTorchFunctionSubclass():
+            return func(*args, **kwargs)
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):  # type: ignore[no-untyped-def]
+        kwargs = {} if kwargs is None else kwargs
+
+        if hasattr(cls, "_DISPATCH_TABLE") and func in cls._DISPATCH_TABLE:
+            return cls._DISPATCH_TABLE[func](func, types, args, kwargs)
+
+        raise NotImplementedError(
+            f"{cls.__name__} dispatch: attempting to run unimplemented operator/function: {func}"
+        )
+
+
+######################################################
+# LayoutType and Layout Tensor Subclass Registration #
+######################################################
+
+
+class PlainMyDTypeLayout(MyDTypeLayout):
+    def __new__(  # type: ignore[no-untyped-def]
+        cls,
+        int_data: torch.Tensor,
+        scale: torch.Tensor,
+        layout_type: LayoutType,
+    ):
+        kwargs = {}
+        kwargs["device"] = int_data.device
+        kwargs["layout"] = kwargs.get("layout") if kwargs.get("layout", False) else int_data.layout  # type: ignore[assignment]
+        kwargs["dtype"] = int_data.dtype  # type: ignore[assignment]
+        kwargs["requires_grad"] = False  # type: ignore[assignment]
+        shape = int_data.shape
+        return torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(
+        self,
+        int_data: torch.Tensor,
+        scale: torch.Tensor,
+        layout_type: LayoutType,
+    ):
+        self.int_data = int_data
+        self.scale = scale
+        self.layout_type = layout_type
+
+    def __tensor_flatten__(self):  # type: ignore[no-untyped-def]
+        return ["int_data", "scale"], [self.layout_type]
+
+    @classmethod
+    def __tensor_unflatten__(cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride):  # type: ignore[no-untyped-def]
+        int_data, scale = tensor_data_dict["int_data"], tensor_data_dict["scale"]
+        (layout_type,) = tensor_attributes
+        return cls(int_data, scale, layout_type)
+
+    @classmethod
+    def from_plain(
+        cls,
+        int_data: torch.Tensor,
+        scale: torch.Tensor,
+        layout_type: LayoutType,
+    ) -> torch.Tensor:
+        """Construct a layout tensor from plain tensors and a layout_type, which main contain
+        extra metadata for packing etc.
+        """
+        assert isinstance(layout_type, PlainLayoutType)
+        return cls(int_data, scale, layout_type)
+
+    def _apply_fn_to_data(self, fn: Callable) -> torch.Tensor:
+        return self.__class__(
+            fn(self.int_data),
+            fn(self.scale),
+            self.layout_type,
+        )
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):  # type: ignore[no-untyped-def]
+        kwargs = {} if kwargs is None else kwargs
+
+        if func is aten.detach.default:
```

Comment:
tbh if we want this subclass to be "best practices" that we are ok with people copy-pasting, I would prefer that we follow the registration pattern that float8 uses: https://github.com/pytorch/ao/blob/main/torchao/float8/float8_tensor.py#L359

Same goes for the NotImplemented stuff here: https://github.com/pytorch/ao/blob/main/torchao/float8/float8_tensor.py#L355

Reply:
we actually have some helper functions in torchao: https://github.com/pytorch/ao/blob/f5703b07acc683653556d04ef970709ba47dba10/torchao/utils.py#L330-L389

so this is not a best-practices example for people to copy pasting actually, the best practices example will be here: https://github.com/pytorch/ao/blob/main/tutorials/developer_api_guide/my_dtype_tensor_subclass.py, also we plan to add a few more things like training, tensor parallelism, maybe FSDP there as well